### ISSUE TITLE
Some questions about reformer implementation

### ISSUE THREAD

**Issue by ag1988** [lucidrains/reformer-pytorch#9]
Hey @lucidrains - thanks for your amazing implementation - it's super cool - I learned so much from it!! I just had the following questions about reformer_pytorch/reformer_pytorch.py -
1)  **line 30**:  I think it should be c.sum() instead of c.sum(dim=-1). Otherwise wouldn't chunked_sum(tensor) just be tensor.sum(dim=-1)? 
2)  **line 72**: In class Chunk(nn.Module), I understand that the intermediate large-dimentional activations of FeedForward Layer will not be stored in the memory due to the use of no_grad() in forward pass of G in the residual block. But I dont see why the **large intermediate gradients** will not be stored during G.backward()? Shouldn't the Chunk module have a custom forward() and backward() in which chunking is also done in backward() with no_grad() so that intermediate large-dimensional gradients of previous chunk are not stored in memory?
3) **line 112**: In hash_vectors() shouldn't one use make_unit_length() to make vecs unit length as line 192 says "_Hashing operates on unit-length vectors..._".
4) **line 169**: ticker = torch.arange(self.n_hashes * seqlen, device=device).unsqueeze(0). Can we expand ticker to buckets.shape ? Maybe its not required due to broadcasting, but can make the code more clear.
5) line **273** : A comment describing why UnsortLogits class is necessary would've been great. But maybe I am requesting this only because this custom backprop stuff is new to me :).

Thanks again for your work,
Ankit

**Comment by lucidrains** [comment]
Hi Ankit! Thank you for the kind words. This has been a learning experience for me as well, as most of the code was faithfully reproduced from the original Trax repository. I will do my best to address your questions.

1. Yes, that is exactly what was intended. I am processing the summation in batches, and then catting them together in the end. This was required for a costly function to deduplicate qk attention values that I reproduced from their repo. I later learned that the authors do not have hard numbers on whether this brings value or not, and I have since turned it off by default after learning the authors frequently have this disabled.

2. If on the forward pass, I process the costly feedforward piece by piece, shouldn't the backward do the same with the gradients? Please correct me if I am mistakened.

3. This comment was copied from their repo and I believe it meant "unit-length key vectors were required for LSH based attention".

4. Sure! Let me check the code later tonight, and I'll make the change if it makes things clearer. Broadcasting confuses me too lol

5. Custom backprop confuses me too :) I faithfully reproduced this from https://github.com/google/trax/blob/master/trax/layers/research/efficient_attention.py#L1013 From what I understand, gradients cannot flow throw indexing operations, which is needed for unsorting the logits. So you connect the gradients by undoing the sort with the indexes you kept around from above.

Phil

**Comment by ag1988** [comment]
1 & 5 : Thanks for the detailed clarifications Phil. I now understand.
2. I have no idea how aggresively pytorch clears the gradients computed at the previous layers but I trust you that it does. In any case, its easy to test this and I'll let you know if there's a problem.

Regards,
Ankit

**Comment by lucidrains** [comment]
Yes, please keep me updated on #2!
